tools/codegen/selective_build/selector.py

```diff
@@ -0,0 +1,148 @@
+    def get_nop_selector() -> 'SelectiveBuilder':
+        return SelectiveBuilder.from_yaml_dict({'include_all_operators': True})
+
+    @staticmethod
+    def from_yaml_dict(data: Dict[str, object]) -> 'SelectiveBuilder':
+        valid_top_level_keys = {
+            'include_all_operators',
+            'debug_info',
+            'operators',
+        }
+        top_level_keys = set(data.keys())
+        if len(top_level_keys - valid_top_level_keys) > 0:
+            raise Exception("Got unexpected top level keys: {}".format(
+                ",".join(top_level_keys - valid_top_level_keys),
+            ))
+        include_all_operators = data.get('include_all_operators', False)
+        assert isinstance(include_all_operators, bool)
+
+        debug_info = None
+        if 'debug_info' in data:
+            di_list = data['debug_info']
+            assert isinstance(di_list, list)
+
+            debug_info = list(map(lambda x: str(x), di_list))
+
+        operators = {}
+        operators_dict = data.get('operators', {})
+        assert isinstance(operators_dict, dict)
+
+        for (k, v) in operators_dict.items():
+            operators[k] = SelectiveBuildOperator.from_yaml_dict(k, v)
+        return SelectiveBuilder(include_all_operators, debug_info, operators)
+
+    @staticmethod
+    def from_yaml_str(config_contents: str) -> 'SelectiveBuilder':
+        contents = yaml.load(config_contents)
+        return SelectiveBuilder.from_yaml_dict(contents)
+
+    @staticmethod
+    def from_yaml_path(config_path: str) -> 'SelectiveBuilder':
+        with open(config_path, 'r') as f:
+            contents = yaml.load(f)
+            return SelectiveBuilder.from_yaml_dict(contents)
+
+    @staticmethod
+    def from_legacy_op_registration_allow_list(
+            allow_list: Set[str],
+            is_root_operator: bool,
+            is_used_for_training: bool) -> 'SelectiveBuilder':
+        operators = {}
+        for op in allow_list:
+            operators[op] = {
+                'name': op,
+                'is_root_operator': is_root_operator,
+                'is_used_for_training': is_used_for_training,
+                'include_all_overloads': True,
+            }
+        return SelectiveBuilder.from_yaml_dict({
+            'operators': operators,
+        })
+
+    def is_operator_selected(self, name: str) -> bool:
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            return True
+        name = strip_operator_overload_name(name)
+        return name in self.operators and self.operators[name].include_all_overloads
+
+    def is_operator_selected_for_training(self, name: str) -> bool:
+        if not self.is_operator_selected(name):
+            return False
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            op: SelectiveBuildOperator = self.operators[name]
+            return op.is_used_for_training
```

Comment:
Is it possible to have the following setup?
entry 1: {name: "aten::add.Tensor", is_root_operator: true, is_used_for_training: false, include_all_overloads: false}
entry 2: {name: "aten::add", is_root_operator: false, is_used_for_training: true, include_all_overloads: true}

Entry 1 is from root operator list of a non-training use case. Entry 2 is from the transitive closure of a training use case.

With the current logic, Entry 1 takes precedence, thus it returns false; but it should be true, right?

Reply:
@ljk53 Good point! I had overlooked this one - will add it to the unit test to make sure that this doesn't regress.
My assumption was that with Training ops, the concept of the transitive closure is non-functional, and we need to always enumerate the complete set of training ops, but that doesn't seem to be true.

I'm guessing this special case applies only to training operator check, and not for the check for `is_root_operator`.

Reply:
Resolved in latest rev (with tests).